### PR TITLE
xapi/xeval: support LIKE

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

File: xapi/xeval/eval.go [hunk 1]
```diff
@@ -319,3 +323,172 @@ func (e *Evaluator) evalTwoChildren(expr *tipb.Expr) (left, right types.Datum, e
...
+}
+
+const (
+	patMatch = 1
+	patOne   = 2
+	patAny   = 3
+)
+
+// handle escapes and wild cards convert pattern characters and pattern types,
+func compilePattern(pattern string, escape byte) (patChars, patTypes []byte) {
```
Comment (BusyJay) [inline xapi/xeval/eval.go]: I think this is too complicated, I propose to use regex instead.

File: xapi/xeval/eval_test.go [hunk 2]
```diff
@@ -261,3 +273,99 @@ func columnExpr(columnID int64) *tipb.Expr {
...
+	switch x := value.(type) {
+	case types.Datum:
+		expr.Children = []*tipb.Expr{datumExpr(x)}
+	case *tipb.Expr:
+		expr.Children = []*tipb.Expr{x}
+	}
+	return expr
+}
+
+func (s *testEvalSuite) TestLike(c *C) {
```
Comment (hanfei19910905) [inline xapi/xeval/eval_test.go]: will it support utf-8 ?

Comment (coocood) [inline xapi/xeval/eval_test.go]: Not fully supported.

File: xapi/xeval/eval.go (hunk 1 above)
Comment (coocood) [inline xapi/xeval/eval.go]: ok

**Comment by coocood** [comment]
@BusyJay PTAL

File: executor/executor_xapi.go [hunk 3]
```diff
@@ -552,6 +552,10 @@ func exprToPBExpr(client kv.Client, expr ast.ExprNode, tn *ast.TableName) *tipb.
 		return binopToPBExpr(client, x, tn)
 	case *ast.ParenthesesExpr:
 		return exprToPBExpr(client, x.Expr, tn)
+	case *ast.PatternLikeExpr:
+		return likeToPBExpr(client, x, tn)
+	case *ast.UnaryOperationExpr:
```
Comment (shenli) [inline executor/executor_xapi.go]: LikeExpr need UnaryOperationExpr?

File: xapi/xeval/eval_test.go (hunk 2 above)
Comment (shenli) [inline xapi/xeval/eval_test.go]: "select name from t where name like '%周煜行%' ;" is OK?

Comment (coocood) [inline xapi/xeval/eval_test.go]: yes

File: executor/executor_xapi.go (hunk 3 above)
Comment (coocood) [inline executor/executor_xapi.go]: Like has `Not` property so we added UnaryOperationExpr for 'Not'.

File: executor/executor_xapi.go [hunk 4]
```diff
@@ -658,3 +662,50 @@ func binopToPBExpr(client kv.Client, expr *ast.BinaryOperationExpr, tn *ast.Tabl
 	}
 	return &tipb.Expr{Tp: tp.Enum(), Children: []*tipb.Expr{leftExpr, rightExpr}}
 }
+
+func likeToPBExpr(client kv.Client, expr *ast.PatternLikeExpr, tn *ast.TableName) *tipb.Expr {
```
Comment (shenli) [inline executor/executor_xapi.go]: Please add some comments to indicate that we only support "xx", "%xx", "xx%", "%xx%" patterns.

**Comment by coocood** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@hanfei19910905 @zxylvlp @zimulala PTAL

File: xapi/xeval/eval.go [hunk 5]
```diff
@@ -319,3 +324,103 @@ func (e *Evaluator) evalTwoChildren(expr *tipb.Expr) (left, right types.Datum, e
...
+		}
+		if last == '%' {
+			return matchPrefix, pattern[:len(pattern)-1]
+		}
+		return matchExact, pattern
+	}
+}
+
+const (
+	matchExact  = 1
```
Comment (zxylvlp) [inline xapi/xeval/eval.go]: Why not iota?

Comment (coocood) [inline xapi/xeval/eval.go]: Just preference, I think either way is ok.

**Comment by zxylvlp** [comment]
LGTM

**Pull request merged**